servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/DefaultHost.java

```diff
@@ -640,4 +647,69 @@ public String toString() {
                     '}';
         }
     }
+
+    private static final class InstrumentedConnectionFactory<Addr, C extends LoadBalancedConnection>
+            extends DelegatingConnectionFactory<Addr, C> {
+
+        private final ConnectTracker connectTracker;
+
+        InstrumentedConnectionFactory(final ConnectionFactory<Addr, C> delegate, ConnectTracker connectTracker) {
+            super(delegate);
+            this.connectTracker = connectTracker;
+        }
+
+        @Override
+        public Single<C> newConnection(Addr addr, @Nullable ContextMap context, @Nullable TransportObserver observer) {
+            return Single.defer(() -> {
+                final long connectStartTime = connectTracker.beforeConnectStart();
+                return delegate().newConnection(addr, context, observer)
+                        .beforeFinally(new ConnectSignalConsumer<>(connectStartTime, connectTracker));
+            });
+        }
+    }
+
+    private static class ConnectSignalConsumer<C extends LoadBalancedConnection> implements TerminalSignalConsumer {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ConnectSignalConsumer> onceUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ConnectSignalConsumer.class, "once");
+
+        private final ConnectTracker connectTracker;
+        private final long connectStartTime;
+        @SuppressWarnings("unused")
+        private volatile int once;
+
+        ConnectSignalConsumer(final long connectStartTime, final ConnectTracker connectTracker) {
+            this.connectStartTime = connectStartTime;
+            this.connectTracker = connectTracker;
+        }
+
+        @Override
+        public void onComplete() {
+            if (once()) {
```

Comment:
beforeFinally operator already guarantees that only one callback will be invoked.
my comment for using filed updated instead of an atomic variable was in general, for any future cases

Reply:
Ah, thank you. Addressed.